### DISCUSSION TITLE
React Hook Form with Material UI and Cleave

### DISCUSSION THREAD

**Discussion by jonyw4** [react-hook-form/react-hook-form#2311] [Q&A] [answered]
Guys, I tried to use React Hook Form with Material UI and Cleave but is always telling me: 

``` text
Material-UI: You have provided a `inputComponent` to the input component
that does not correctly handle the `inputRef` prop.
Make sure the `inputRef` prop is called with a HTMLInputElement.
```
What I am doing wrong? Here is an example:
https://codesandbox.io/s/rhf-mui-cleave-ult8d?file=/src/App.js

**Comment by bluebill1049** [comment]
you mean `inputProps` not **I**nputProps right?

**Reply by jonyw4** [reply]
No, `InputProps` is for `TextField` and `inputProps` is for `Input`. But it's not working both in `TextField` and `Input` component from Material-UI. I don't know what I am doing wrong...
If I delete the `ref={inputRef}` part from the `Cleave` component the Material-UI doesn't complain, but the input starts to behave weird. The label goes down, sometimes lose focus. I believe that I am missing something....

![image](https://user-images.githubusercontent.com/14056669/88175775-b65d8200-cbfc-11ea-85d0-47b3f0e4c443.png)

**Reply by bluebill1049** [reply]
first time i have seen a prop name is uppercase. will take a look again.

**Reply by jonyw4** [reply]
Yeah, the documentation of text field it's in here:
https://material-ui.com/api/text-field/

**Reply by jonyw4** [reply]
I found[ in the Cleave documentation](https://github.com/nosir/cleave.js/blob/master/doc/reactjs-component-usage.md#how-to-get-ref-of-the-input-field) that ref must be passed in `htmlRef` prop. This solves the main problem, but I am having another problem. When the state of the form changes / the components rerender, the input lose focus. Do you know why?
I update the codesanbox file

**Reply by bluebill1049** [reply]
hey  @jonyw4 why i couldn't see the error/warning in your CSB.

**Reply by bluebill1049** [reply]
<img width="1238" alt="Screen Shot 2020-07-23 at 3 07 13 pm" src="https://user-images.githubusercontent.com/10513364/88253709-3287c080-ccf6-11ea-9dba-54e17561ae2e.png">

**Reply by jonyw4** [reply]
Yea, that its what I said in the previous comment 😄.
I could solve the main problem reading the documentation from Cleave. The `ref` should be added in prop `htmlRef`. This stop the warning.
But now sometimes the field lose focus.  I don't know if this is a problem with react-hook-form or maybe cleave or material, I am investigating.
After you enter the first character the field lose focus. This happens again if you erase all the content of the field and try to enter again the first character. I think that its related with `ref`...

**Comment by bluebill1049** [comment]
> Yea, that its what I said in the previous comment 😄.
> I could solve the main problem reading the documentation from Cleave. The `ref` should be added in prop `htmlRef`. This stop the warning.
> But now sometimes the field lose focus. I don't know if this is a problem with react-hook-form or maybe cleave or material, I am investigating.
> After you enter the first character the field lose focus. This happens again if you erase all the content of the field and try to enter again the first character. I think that its related with `ref`...

oops, didn't read careful enough, i will take a look again, i think it's related to ref as well.

**Reply by jonyw4** [reply]
Thanks Bill!

**Reply by bluebill1049** [reply]
what about this approach? https://codesandbox.io/s/rhf-mui-cleave-5lc7l?file=/src/App.js

otherwise, you have to use custom register.

**Reply by jonyw4** [reply]
I will check, thanks for the help!
What is the difference of using `render` or `as`?

**Reply by bluebill1049** [reply]
Take a look here: http://localhost:8000/api#Controller

**Reply by jonyw4** [reply]
This solve the issue. Thanks Bill!!!

**Reply by bluebill1049** [reply]
awesome!

**Comment by thiagofernando** [comment]
Do you have any example with OutlinedInput element?

**Reply by thiagofernando** [reply]
@bluebill1049   there is a best approach? 
https://codesandbox.io/s/rhf-mui-cleave-d8r26

doesnt work in all mode (blur)

**Reply by jonyw4** [reply]
I think that I fix your demo:
https://codesandbox.io/s/rhf-mui-cleave-26jjg

Take a look in my answer: https://github.com/react-hook-form/react-hook-form/discussions/2311#discussioncomment-46086

**Comment by jonyw4** [comment, marked as answer]
@bluebill1049  @thiagofernando 
Initially, the solution that @bluebill1049 gave to me work, but when I try to send the form didn't. So today I found the solution:
1. Pass all props from React Hook Form `Controller`  `render` to `TextField`
2. Create a component for Cleave that pass the `inputRef` of `TextField` to `htmlRef` of Cleave
3. (Optional) Use the prop  `inputProps`  of `TextField` to define the options of `Cleave`

Here a demo:
https://codesandbox.io/s/rhf-mui-cleave-9uqxc?file=/src/App.js:1176-1186

Now its officially resolved!

**Reply by bluebill1049** [reply]
thanks @jonyw4!

**Reply by getsetbro** [reply]
Looks good. I am trying to get it to remain invalid if only spaces are entered. It works but when Cleave adds mask items the validator does not trim() those.

**Reply by Zoot01** [reply]
currently working in @mui/material^5.11.6, cleave.js^1.6.0, react-hook-form^7.43.0

```typescript
const CleaveInput = forwardRef(({ ...props }: any, ref) => {
  const { options, inputRef, ...other } = props;
  return <Cleave {...other} ref={inputRef} options={options} />;
});

// component to return 
<FormControl fullWidth>
                  <Controller
                    name="phoneNumber"
                    control={accountControl}
                    rules={{ required: true }}
                    render={({ field: { value, onChange } }) => (
                      <TextField
                        type="text"
                        value={value}
                        label="Phone number"
                        onChange={onChange}
                        error={Boolean(accountErrors.phoneNumber)}
                        placeholder="234 567 8901"
                        aria-describedby="stepper-linear-account-phoneNumber"
                        inputProps={{
                          options: { phone: true, phoneRegionCode: 'US' },
                        }}
                        InputProps={{
                          inputComponent: CleaveInput,
                        }}
                      />
                    )}
                  />
                  {accountErrors.phoneNumber && (
                    <FormHelperText
                      sx={{ color: 'error.main' }}
                      id="stepper-linear-account-phoneNumber"
                    >
                      {accountErrors.phoneNumber.message}
                    </FormHelperText>
                  )}
</FormControl>

```

**Comment by thiagofernando** [comment]
@bluebill1049 @jonyw4  i change the code to explicit my point. Basically the difference between this 2 is 


https://codesandbox.io/s/rhf-mui-cleave-d8r26?file=/src/App.js

![rhf cleave](https://user-images.githubusercontent.com/11356440/89279739-1a7d3e80-d61e-11ea-8d96-34a1609f0f32.gif)

i would like to have the same behavior like no cleave option

**Reply by jonyw4** [reply]
Hey Thiago, here is the example correctly:
https://codesandbox.io/s/rhf-mui-cleave-co545

I see that your code doesn't use the recent resolution that I post here, [take a look](https://github.com/react-hook-form/react-hook-form/discussions/2311#discussioncomment-46086). So here some tips that I found in this proccess:

1. You don't need to use `state` or `setValues` to use react-hook-form to change the values of the form. You can use the [Controller ](https://react-hook-form.com/api#Controller) to do that, it's so much easier and cleaner. You can use it in many ways, take a look in the docs and in the [examples](https://codesandbox.io/s/react-hook-form-controller-079xx).
2. [When you use the render prop from Controller like here](https://codesandbox.io/s/rhf-mui-cleave-d8r26?file=/src/App.js:1737-1755) you should map all these props ( `onChange`, `onBlur` and `value`) to the component that you're trying to implement. In this case, all Material UI inputs are equally equivalent with the implementation of these props,[ so you can spread all then, like I did here](https://codesandbox.io/s/rhf-mui-cleave-co545?file=/src/App.js:1592-1602). 
3. You must set the `inputRef` from Material UI inputs to the prop `htmlRef` of Cleave.  [Take a look in the Cleave docs](https://github.com/nosir/cleave.js/blob/master/doc/reactjs-component-usage.md#how-to-get-ref-of-the-input-field)
4. When using Material UI with React Hook Form you must set a to `defaultValue=""` to the Controller. If don't the Material UI will warn you that you are changing from uncontrolled to controlled input.

**Reply by thiagofernando** [reply]
Thanks for your advices about state, but i think you dont understood the problem.  I already send all props to componet, but with Controller and Cleave, when i click in the field, and click out, the error trigger is not activated like in No Controller approach. Please see the difference in GIF

one more question @jonyw4 how i get the cleave rawValue without a state?

**Reply by jonyw4** [reply]
I understand now what you want, but I think that it's not a Material UI or either a Cleave problem. Its supposed to work like that... See [another examples here](https://github.com/react-hook-form/react-hook-form/tree/master/examples)
And in your example you don't using the [Controller component to control the input](https://react-hook-form.com/api#Controller), so that I guess that it's why your input is working different.

About the rawValue of [Cleave you can read more here](https://github.com/nosir/cleave.js/blob/master/doc/reactjs-component-usage.md#how-does-it-work), but basically you can access the event of Cleave onChange.

I update the file using Yup as validation and log the rawValue. Works fine! You can see [when the validation is triggered as well in the docs](https://react-hook-form.com/api#useForm). 
https://codesandbox.io/s/rhf-mui-cleave-co545

**Reply by thiagofernando** [reply]
@jonyw4 @bluebill1049  the cep2 works just like i want with Material Ui and Rhf. The problem is when I use controller
